vllm/v1/core/kv_cache_manager.py

```diff
@@ -98,10 +107,17 @@ def get_computed_blocks(self, request: Request) -> List[KVCacheBlock]:
             # not computed yet for sure.
             if cached_block := self._get_cached_block(block_hash):
                 computed_blocks.append(cached_block)
+                num_free_blocks += cached_block.ref_cnt == 0
             else:
                 break
 
-        return computed_blocks
+        # Remove the last computed block if all blocks are computed.
+        num_new_tokens = (request.num_tokens -
+                          len(computed_blocks) * self.block_size)
+        if num_new_tokens == 0:
+            num_free_blocks -= computed_blocks.pop().ref_cnt == 0
```

Comment:
Can we handle this in scheduler, rather than in this method?

I think this constraint comes from the scheduler, and here we are coupling the scheduler's constraint with the KV cache manager.
If we move it back to the scheduler, we can make the `get_computed_blocks` method a function that is only concerned on cache hit, which is simple and clean.